storage/database/db_manager.go

```diff
@@ -371,10 +387,85 @@ func (dbm *databaseManager) IsParallelDBWrite() bool {
 	return dbm.config.ParallelDBWrite
 }
 
+func (dbm *databaseManager) InMigration() bool {
+	return dbm.inMigration
+}
+
 func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
+	if dbEntryType == StateTrieDB && dbm.inMigration {
+		return dbm.newStateTrieDB.NewBatch()
+	}
 	return dbm.getDatabase(dbEntryType).NewBatch()
 }
 
+func (dbm *databaseManager) getStateTrieMigrationInfo() (bool, string) {
+	miscDB := dbm.getDatabase(MiscDB)
+
+	enc, _ := miscDB.Get(migrationStatusKey)
+	if len(enc) == 0 {
+		return false, ""
+	}
+
+	var status uint64
+	if err := rlp.DecodeBytes(enc, &status); err != nil {
+		logger.Error("Failed to decode migration status", "err", err)
+		return false, ""
+	}
+
+	if status == 0 {
+		return false, ""
+	}
+
+	enc, _ = miscDB.Get(databaseDirKey(uint64(StateTrieDB)))
+	return true, string(enc)
+}
+
+func createStateTrieMigrationDB(dbc *DBConfig, newDBDir string) Database {
+	oldDBConfig := dbc
+	newDBConfig := getDBEntryConfig(oldDBConfig, StateTrieDB)
+	newDBConfig.Dir = newDBDir
+	var newDB Database
+	var err error
+	if newDBConfig.NumStateTriePartitions > 1 {
+		newDB, err = newPartitionedDB(newDBConfig, StateTrieDB, newDBConfig.NumStateTriePartitions)
+	} else {
+		newDB, err = newDatabase(newDBConfig, StateTrieDB)
+	}
+	if err != nil {
+		logger.Crit("Failed to create a new database for state trie migration", "err", err)
+	}
+	logger.Info("Created a new database for state trie migration", "newStateTrieDB", newDBConfig.Dir)
+	return newDB
+}
+
+func (dbm *databaseManager) SetStateTrieMigrationDB(blockNum uint64) {
+	if !dbm.config.Partitioned {
+		logger.Warn("Setting a new database for state trie migration is allowed for partitioned database only")
+		return
+	}
+
+	logger.Info("Start setting a new database for state trie migration", "blockNum", blockNum)
+
+	// Store the directory of the new database first to avoid missing directory name situation
+	newDBDir := dbm.config.Dir + strconv.FormatUint(blockNum, 10)
+	miscDB := dbm.getDatabase(MiscDB)
+	if err := miscDB.Put(databaseDirKey(uint64(StateTrieDB)), []byte(newDBDir)); err != nil {
+		logger.Crit("Failed to store a new database directory", "newDBDir", newDBDir, "err", err)
+	}
+
+	// After storing the directory, create a new database for migration process.
+	newDB := createStateTrieMigrationDB(dbm.config, newDBDir)
+
+	// After creating the database, store migration status to the database.
+	dbm.newStateTrieDB = newDB
+	dbm.inMigration = true
+	if err := miscDB.Put(migrationStatusKey, encodeUint64(1)); err != nil {
+		logger.Crit("Failed to store the migration status", "status", 1, "err", err)
```

Comment:
If we failed here, it would be better to remove the created database `newDBDir` for the next trial.

Reply:
Updated!